instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/handlers.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require_relative 'mappers/attribute'
+require_relative 'handlers/default'
+require_relative 'handlers/enqueue'
+require_relative 'handlers/perform'
+
+module OpenTelemetry
+  module Instrumentation
+    module ActiveJob
+      # Module that contains custom event handlers, which are used to generate spans per event
+      module Handlers
+        module_function
+
+        # Subscribes Event Handlers to relevant ActiveJob notifications
+        #
+        # The following events are recorded as spans:
+        # - enqueue
+        # - enqueue_at
+        # - enqueue_retry
+        # - perform
+        # - retry_stopped
+        # - discard
+        #
+        # Ingress and Egress spans (perform, enqueue, enqueue_at) use Messaging semantic conventions for naming the span,
+        # while internal spans keep their ActiveSupport event name.
+        #
+        # @note this method is not thread safe and should not be used in a multi-threaded context
+        # @note Why no perform_start?
+        #       This event causes much heartache as it is the first in a series of events that is triggered.
+        #       It should not be the ingress span because it does not measure anything.
+        #       https://github.com/rails/rails/blob/v6.1.7.6/activejob/lib/active_job/instrumentation.rb#L14
+        #       https://github.com/rails/rails/blob/v7.0.8/activejob/lib/active_job/instrumentation.rb#L19
+        def subscribe
+          return unless Array(@subscriptions).empty?
+
+          tracer = OpenTelemetry.tracer_provider.tracer(ActiveJob.name, ActiveJob::VERSION)
+          mapper = Mappers::Attribute.new
+          config = ActiveJob::Instrumentation.instance.config
+          parent_span_provider = OpenTelemetry::Instrumentation::ActiveJob
+
+          # TODO, use delegation instead of inheritance
+          default_handler = Handlers::Default.new(tracer, parent_span_provider, mapper, config)
+          enqueue_handler = Handlers::Enqueue.new(tracer, parent_span_provider, mapper, config)
+          perform_handler = Handlers::Perform.new(tracer, parent_span_provider, mapper, config)
```

Comment:
Our general pattern in instrumentation is to always reach for the `tracer` from the instrumentation _when we need it_ rather than caching it in various helpers. I think the deviation from that code pattern here is the trigger for https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/688

Reply:
Yessir! This was intentional on my part in the hopes of starting a new tread 😄 

I wanted to use dependency injection for everything that the handlers needed instead of using a service locator or a mix of the two. 

Though we are hashing out some details in #688, do you think having this asymmetry is problematic?

Reply:
@fbogsany did you see my comment about this in the other issue?

Is this a deal breaker?

Reply:
Given the results of the conversation in #688, I have changed the handlers to use the service locator pattern and look up the tracer from the Instrumentation singleton to creating spans.